core/src/main/scala/caliban/ws/WebSocketHooks.scala

```diff
@@ -0,0 +1,125 @@
+package caliban.ws
+
+import caliban.{ GraphQLWSOutput, InputValue, ResponseValue }
+import zio.ZIO
+import zio.stream.ZStream
+
+trait StreamTransformer[-R, +E] {
+  def transform[R1 <: R, E1 >: E](stream: ZStream[R1, E1, GraphQLWSOutput]): ZStream[R1, E1, GraphQLWSOutput]
```

Comment:
I’m wondering if we could use a `ZPipeline` for this instead of defining a new type?

Reply:
Yeah I think that would be ideal, although my only concern is that it's a user-facing breaking change.

Having said that, the migration should be very straightforward by simply using `ZPipeline.fromFunction` so perhaps we should just do it and document the migration

Reply:
This is already breaking right? So as long as we are doing that can probably also move this

Reply:
It's not bin-compatible, but other than a deprecation warning the previous changes were source compatible.

Actually I think we might be able to keep it source-compatible (at least for the deprecated trait). Let me give it a go

Reply:
Ok I overloaded the `message` method in the deprecated `WebSocketHooks` object to accept a `StreamTransformer`, that way it should be source-compatible for users that don't manually create an instance of `WebSocketHooks`